apis/src/components/organisms/analysis/tree.rs

```diff
@@ -0,0 +1,130 @@
+        for (i, m) in gs.state.history.moves.iter().enumerate() {
+            let parent_id = if i == 0 { None } else { Some(i - 1) };
+            let node = TreeNode {
+                turn: i,
+                piece: m.0.clone(),
+                position: m.1.clone(),
+            };
+            tree.add_node(Node::new(i, Some(node)), parent_id.as_ref())
+                .unwrap();
+        }
+        let node = if !gs.state.history.moves.is_empty() {
+            Some(gs.state.history.moves.len() - 1)
+        } else {
+            None
+        };
+        Self {
+            current_node: node,
+            tree,
+        }
+    }
+    pub fn update_node(&mut self, node: usize) {
+        let mut moves: VecDeque<(String, String)> = VecDeque::new();
+        let mut tmp_node = node;
+        loop {
+            let node = self.tree.get_node(&tmp_node).unwrap().get_value().unwrap();
+            moves.push_front((node.piece.clone(), node.position.clone()));
+            if let Some(parent) = self.tree.get_node(&tmp_node).unwrap().get_parent() {
+                tmp_node = parent
+            } else {
+                break;
+            }
+        }
+        let mut new_history = History::new();
+        new_history.game_type = GameType::MLP;
+        new_history.moves = moves.into();
+        let state = State::new_from_history(&new_history).expect("Got state from current branch");
+        self.current_node = Some(node);
+        let game_state = expect_context::<GameStateSignal>();
+        let history_turn = if game_state
+            .signal
+            .get_untracked()
+            .state
+            .history
+            .moves
+            .is_empty()
+        {
+            None
+        } else {
+            Some(game_state.signal.get_untracked().state.history.moves.len() - 1)
+        };
+        game_state
+            .signal
+            .update(|gs: &mut crate::providers::game_state::GameState| {
+                gs.state = state;
+                gs.history_turn = history_turn;
+                gs.move_info.reset();
+            });
+    }
+    pub fn add_node(&mut self, last_move: (String, String)) {
+        let node_id = self.tree.get_nodes().len();
+        let turn = if let Some(curr_node) = self.current_node {
+            1 + self
+                .tree
+                .get_node(&curr_node)
+                .unwrap()
+                .get_value()
+                .unwrap()
+                .turn
+        } else {
+            0
+        };
+        let node = TreeNode {
+            turn,
+            piece: last_move.0.clone(),
+            position: last_move.1.clone(),
+        };
+        self.tree
+            .add_node(Node::new(node_id, Some(node)), self.current_node.as_ref())
+            .unwrap();
+        self.current_node = Some(node_id);
+    }
+
+    pub fn in_path(&self) -> Vec<bool> {
+        let mut in_path = vec![false; self.tree.get_nodes().len() + 5];
```

Comment:
What happens in this function?

Reply:
returned a boolean vec with true if the node is is the path, its not needed though